### ISSUE TITLE
Addons - EKS module FORCES to downgrade VPC-CNI addon version that was upgraded from AWS Management Console

### ISSUE THREAD

**Issue by luarx** [terraform-aws-modules/terraform-aws-eks#2758]
## Description

Terraform EKS module **forces to downgrade VPC-CNI addon version** which was upgraded from the AWS Management 
 Console

## ⚠️ Note

Before you submit an issue, please perform the following first:

1. Remove the local `.terraform` directory (! ONLY if state is stored remotely, which hopefully you are following that best practice!): `rm -rf .terraform/`
2. Re-initialize the project root to pull down modules: `terraform init`
3. Re-attempt your terraform plan or apply and check if the issue still persists

**DONE! The issue persists**

## Versions

- Module version [Required]: 19.16.0

- Terraform version: 1.5.7

- Provider version(s): 
```
+ provider registry.terraform.io/hashicorp/aws v4.57.1
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.2
+ provider registry.terraform.io/hashicorp/kubernetes v2.23.0
+ provider registry.terraform.io/hashicorp/local v2.4.0
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/random v3.5.1
+ provider registry.terraform.io/hashicorp/time v0.9.1
+ provider registry.terraform.io/hashicorp/tls v4.0.4
```

## Reproduction Code [Required]

```
module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version         = "19.16.0"
  cluster_name    = local.cluster_name
  cluster_version = "1.26"
  cluster_endpoint_public_access = true
  cluster_endpoint_private_access = false
  subnet_ids      = module.vpc.private_subnets

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {
      resolve_conflicts = "OVERWRITE"
    }
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
    }
    aws-ebs-csi-driver = {
      resolve_conflicts = "OVERWRITE"
    }
  }
vpc_id = module.vpc.vpc_id

eks_managed_node_groups = {
    general = {
      min_size     = 0
      max_size     = 8

      instance_types = ["m6i.xlarge"]

      subnet_ids = [module.vpc.private_subnets[0]]
    },
}
}
```

- Are you using workspaces? No
- Have you cleared the local cache (see Notice section above)? Yes
- List steps in order that led up to the issue you encountered:

1. I have updated EKS module and providers
From:
```
+ provider registry.terraform.io/hashicorp/aws v4.23.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.2.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.13.1
+ provider registry.terraform.io/hashicorp/local v2.2.3
+ provider registry.terraform.io/hashicorp/null v3.1.1
+ provider registry.terraform.io/hashicorp/random v3.1.3
+ provider registry.terraform.io/hashicorp/tls v4.0.3
```
To:
```
+ provider registry.terraform.io/hashicorp/aws v4.57.1
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.2
+ provider registry.terraform.io/hashicorp/kubernetes v2.23.0
+ provider registry.terraform.io/hashicorp/local v2.4.0
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/random v3.5.1
+ provider registry.terraform.io/hashicorp/time v0.9.1
+ provider registry.terraform.io/hashicorp/tls v4.0.4
```

2.  I have run terraform apply. The new Terraform EKS module **has applied only some changes** because it complained that it is not possible to downgrade more than 1 VPC-CNI addon version at the same time.
3.  After that, I have run again terraform apply and the terraform output says that it wants to downgrade my current VPC-CNI addon version (which was upgraded some days ago using the AWS Management Console). This is weird because before upgrading, Terraform didn't want to downgrade that version.

## Expected behavior

I want that Terraform EKS module does not consider that I want to downgrade VPC-CNI addon version.

### Terminal Output Screenshot(s)

```
Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply" which may have affected this plan:

  # module.eks.aws_eks_addon.this["vpc-cni"] has changed
  ~ resource "aws_eks_addon" "this" {
      ~ addon_version     = "v1.12.5-eksbuild.2" -> "v1.14.1-eksbuild.1"
        id                = "eks-staging:vpc-cni"
      ~ modified_at       = "2023-09-27T12:18:04Z" -> "2023-09-27T12:44:49Z"
        tags              = {}
        # (6 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }


Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to undo or respond to these
changes.

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.eks.aws_eks_addon.this["vpc-cni"] will be updated in-place
  ~ resource "aws_eks_addon" "this" {
      ~ addon_version     = "v1.14.1-eksbuild.1" -> "v1.12.5-eksbuild.2"
        id                = "eks-staging:vpc-cni"
        tags              = {}
        # (7 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

## Additional context

- It also happens with other addons
- I have a different cluster running EKS Module v18.31.2 (the same that I had in the issue cluster before upgrading) using a newer addon version and it does not force me to downgrade
- I have tried to downgrade from the AWS Management Console the version until match the one that EKS Module wants, run terraform apply (everything was ok), upgrade from the Management  Console again the VPC-CNI addon version, run terraform apply but the EKS module forces again to downgrade the version.

**Comment by bryantbiggs** [comment]
It depends on what you mean by "upgrade" - for starters, its not recommended to make changes to resources Terraform controls outside of Terraform; this will cause conflicts like you are showing above

For the VPC CNI, the default version for 1.26 is `v1.12.5-eksbuild.2` - and when you enable addons via this module on v19.x, the default behavior is to use the addons default version for the given cluster version. Given the output you've shown above, it looks the module is doing the correct thing and selecting the default version of the addon for EKS 1.26

If you wish to use the latest version, you would need to specify [`most_recent = true`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/918aa7cc40cbc072836410747834de64d84f514d/examples/complete/main.tf#L58) for the addon

Do note that EKS addons typically require upgrading one minor version at a time 

```sh
K8S_VERSION='1.26'

for ADDON in vpc-cni; do
  LATEST=$(aws eks describe-addon-versions --region us-east-1 --addon-name ${ADDON} \
    --kubernetes-version $K8S_VERSION --query 'addons[0].addonVersions[0].addonVersion' --output text)
  LIST=$(aws eks describe-addon-versions --region us-east-1 --addon-name ${ADDON} \
    --kubernetes-version $K8S_VERSION)
  DEFAULT=$(echo $LIST | jq -r '.addons[] .addonVersions[] | select(.compatibilities[] .defaultVersion==true).addonVersion')

  echo "${ADDON} latest version for ${K8S_VERSION}: ${LATEST}"
  echo "${ADDON} default version for ${K8S_VERSION}: ${DEFAULT}"
done
```

```text
vpc-cni latest version for 1.26: v1.15.0-eksbuild.2
vpc-cni default version for 1.26: v1.12.5-eksbuild.2
```

**Comment by luarx** [comment]
> It depends on what you mean by "upgrade" - for starters, its not recommended to make changes to resources Terraform controls outside of Terraform; this will cause conflicts like you are showing above

I totally agree with this. Just to clarify that for this specific case it was easier for me to use the AWS Management Console after checking in some ocassions that there were not conflicts with Terraform EKS module when doing that

> For the VPC CNI, the default version for 1.26 is v1.12.5-eksbuild.2 - and when you enable addons via this module on v19.x, the default behavior is to use the addons default version for the given cluster version. Given the output you've shown above, it looks the module is doing the correct thing and selecting the default version of the addon for EKS 1.26

**I didn't know the behaviour in v19, I like that indeed!** With this I will not probably use the AWS Management Console to upgrade EKS addons :)

Thanks for the whole bunch of information, it has helped me to understand the current situation 🙌
This issue is for sure clarified

**Comment by bryantbiggs** [comment]
also, when you manage the addons with this module and do NOT specify a specific version, the module will automatically select the appropriate addon version even when upgrading the EKS version. It uses a [data source](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/918aa7cc40cbc072836410747834de64d84f514d/main.tf#L432-L438) to make calls similar to the script I added above, so when you upgrade to the next EKS version, it will call the addons API and set the appropriate addon version for the given EKS version (either using the default addon version, or the latest addon version)

tl;dr - managing the addons with this module ensures the addon versions are always upgraded and kept compatible with the given EKS version

**Comment by luarx** [comment]
Perfect, thanks for that information! I am more confident having that information when upgrading my cluster 😄 

Now the only thing that I am worried about is that when upgrading the K8s cluster version specified in the module, it seems that Terraform wants to upgrade all the worker node groups at the same time... I haven't tried because I want to upgrade in different times those worker node groups (what is allowed using the AWS Management Console)

**Comment by askulkarni2** [comment]
>upgrade in different times 

@luarx not sure what you meant by that but have you looked at [update_config](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_node_group.html#update_config-configuration-block)? As specified you can set the upgrade batch to a number or a percent of nodes.

**Comment by luarx** [comment]
Oh that is useful @askulkarni2 ! 🙌 
Using that it is smoother the upgrade as there can not be more than X nodes unavailable using [this variable](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/918aa7cc40cbc072836410747834de64d84f514d/modules/eks-managed-node-group/variables.tf#L389)

Although I think that it is not possible to specify that I want to upgrade 1 node group and that the next one can not be upgraded until the first one is not fully upgraded 🤔

**Comment by arbreezy** [comment]
I think it's worth mentioning that upgrading to latest add_on version while your infrastructure is behind at least two versions, it will fail because the process requires an incremental version upgrading, so be careful when using the latest and not running often Terraform plan/apply against your infrastructure
